Fix: Query to select most used tax rule is simply ridicoulous

Codencode opened the pull request:
Added index `shop_tax` to the table `product_shop` and optimized the query of the method `Product::getIdTaxRulesGroupMostUsed()`

<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project!

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/#pull-requests

For type and category see:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#type--category
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.2.x
| Description?      | The query in the Product::getIdTaxRulesGroupMostUsed() method, in the case of a database with many products, makes the insertion of a new product very slow (This was noticed by @matteolavaggi).
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | See https://github.com/PrestaShop/PrestaShop/issues/38095 and see: https://github.com/PrestaShop/PrestaShop/pull/38251#issuecomment-2747196930
| UI Tests          | 
| Fixed issue or discussion?     | Fixes #38095
| Related PRs       | https://github.com/PrestaShop/autoupgrade/pull/1272
| Sponsor company   | @Codencode

Codencode committed:
Added index "shop_tax" to the table "product_shop" and optimized the query of the method Product::getIdTaxRulesGroupMostUsed(

Codencode commented:
N.B.
The change adds an index to the product_shop table, so it should also be added to the autoupgrade module: https://github.com/PrestaShop/autoupgrade.

jolelievre commented:
Ah it seems it has impacts on the behat test, maybe the default tax rules has changed based on your improvement? Does it mean the changes you made use different criteria to select the most used one?

Codencode commented:
Hi @jolelievre,

I don’t think so, the query was suggested by @Hlavtox.

Anyway, I’ll try to look into it further now.

@Hlavtox, what do you think?

Codencode commented:
I'm running some tests and I always get the same result.

The original query is:
```sql
SELECT id_tax_rules_group
FROM (
    SELECT COUNT(*) n, product_shop.id_tax_rules_group
    FROM ps_product p
    INNER JOIN ps_product_shop product_shop
    ON (product_shop.id_product = p.id_product AND product_shop.id_shop = 1)
    JOIN ps_tax_rules_group trg ON (product_shop.id_tax_rules_group = trg.id_tax_rules_group)
    WHERE trg.active = 1 AND trg.deleted = 0
    GROUP BY product_shop.id_tax_rules_group
    ORDER BY n DESC
    LIMIT 1
) most_used;
```


The modified query is:
```sql
SELECT id_tax_rules_group
FROM (
    SELECT DISTINCT product_shop.id_tax_rules_group, COUNT(*) count
    FROM ps_product_shop product_shop
    JOIN ps_tax_rules_group trg ON (product_shop.id_tax_rules_group = trg.id_tax_rules_group)
    WHERE trg.active = 1 AND trg.deleted = 0 AND product_shop.id_shop IN (1)
    ORDER BY count DESC
) most_used;
```

Codencode commented:
Indeed, in the case of a multishop, it yields a different result.

Moreover, the query had been suggested by @Eolia.

The one suggested by @Hlavtox was this, and it seems to work:

```sql
SELECT ps.id_tax_rules_group
FROM ps_product_shop ps
INNER JOIN ps_tax_rules_group trg ON ps.id_tax_rules_group = trg.id_tax_rules_group
WHERE trg.active = 1 AND trg.deleted = 0 AND ps.id_shop IN (2)
GROUP BY ps.id_tax_rules_group
ORDER BY COUNT(*) DESC
```

Now I'll update the PR and let's see what result it gives.

Codencode commented:
@jolelievre with the modification, the tests have been passed, so now it should be okay.

Eolia commented:
> Indeed, in the case of a multishop, it yields a different result.

This is the desired behavior.
If shop A primarily sells products at 20% and shop B at 10%, what we want is the majority rate in the current shop.

Hlavtox reviewed (approved):
I think we can approve, but should be tested in multistore scenario!

Betafer commented:
Thanks, for this contribution many parts of prestashop have similar problems, it's nice that finally someone notices it! thanks @Codencode

Hlavtox commented:
@PrestaShop/committers Guys let's get this merged, it's a nice performance improvement and an amazing job from @Codencode  :-)

Codencode commented:
@Betafer,
I've been using Prestashop since 2010 (since version 1.1), and I've encountered hundreds of issues over the years.
At first, I only made fixes on my clients' sites because I wasn't using GitHub, but for some time now, I've been trying to report issues and—when possible—solve them myself.
I believe Prestashop is an excellent CMS, and in my opinion, at least the GitHub version is improving significantly.

@Hlavtox, thank you very much!

matteolavaggi commented:
I can't definitely accept excellent for a product that after 14 years still
not support at least

Seriously Caching
Elastich Search searching and facets

Really. Also the hook system.... something from 1990 ...
Btw here we are. Add-ons store growth a lot 🤣

Il Mer 26 Mar 2025, 10:44 Codencode ***@***.***> ha scritto:

> @Betafer <https://github.com/Betafer>,
> I've been using Prestashop since 2010 (since version 1.1), and I've
> encountered hundreds of issues over the years.
> At first, I only made fixes on my clients' sites because I wasn't using
> GitHub, but for some time now, I've been trying to report issues and—when
> possible—solve them myself.
> I believe Prestashop is an excellent CMS, and in my opinion, at least the
> GitHub version is improving significantly.
>
> @Hlavtox <https://github.com/Hlavtox>, thank you very much!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PrestaShop/PrestaShop/pull/38251#issuecomment-2753778184>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF5GL32I3MI37K2XLNWL2UT2WJSITAVCNFSM6AAAAABZAD5XMWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONJTG43TQMJYGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: Codencode]*Codencode* left a comment (PrestaShop/PrestaShop#38251)
> <https://github.com/PrestaShop/PrestaShop/pull/38251#issuecomment-2753778184>
>
> @Betafer <https://github.com/Betafer>,
> I've been using Prestashop since 2010 (since version 1.1), and I've
> encountered hundreds of issues over the years.
> At first, I only made fixes on my clients' sites because I wasn't using
> GitHub, but for some time now, I've been trying to report issues and—when
> possible—solve them myself.
> I believe Prestashop is an excellent CMS, and in my opinion, at least the
> GitHub version is improving significantly.
>
> @Hlavtox <https://github.com/Hlavtox>, thank you very much!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PrestaShop/PrestaShop/pull/38251#issuecomment-2753778184>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF5GL32I3MI37K2XLNWL2UT2WJSITAVCNFSM6AAAAABZAD5XMWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONJTG43TQMJYGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Betafer commented:
> I can't definitely accept excellent for a product that after 14 years still not support at least Seriously Caching Elastich Search searching and facets Really. Also the hook system.... something from 1990 ... Btw here we are. Add-ons store growth a lot 🤣 Il Mer 26 Mar 2025, 10:44 Codencode ***@***.***> ha scritto:
> […](#)
> @Betafer <https://github.com/Betafer>, I've been using Prestashop since 2010 (since version 1.1), and I've encountered hundreds of issues over the years. At first, I only made fixes on my clients' sites because I wasn't using GitHub, but for some time now, I've been trying to report issues and—when possible—solve them myself. I believe Prestashop is an excellent CMS, and in my opinion, at least the GitHub version is improving significantly. @Hlavtox <https://github.com/Hlavtox>, thank you very much! — Reply to this email directly, view it on GitHub <[#38251 (comment)](https://github.com/PrestaShop/PrestaShop/pull/38251#issuecomment-2753778184)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AF5GL32I3MI37K2XLNWL2UT2WJSITAVCNFSM6AAAAABZAD5XMWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONJTG43TQMJYGQ> . You are receiving this because you were mentioned.Message ID: ***@***.***> [image: Codencode]*Codencode* left a comment ([PrestaShop/PrestaShop#38251](https://github.com/PrestaShop/PrestaShop/pull/38251)) <[#38251 (comment)](https://github.com/PrestaShop/PrestaShop/pull/38251#issuecomment-2753778184)> @Betafer <https://github.com/Betafer>, I've been using Prestashop since 2010 (since version 1.1), and I've encountered hundreds of issues over the years. At first, I only made fixes on my clients' sites because I wasn't using GitHub, but for some time now, I've been trying to report issues and—when possible—solve them myself. I believe Prestashop is an excellent CMS, and in my opinion, at least the GitHub version is improving significantly. @Hlavtox <https://github.com/Hlavtox>, thank you very much! — Reply to this email directly, view it on GitHub <[#38251 (comment)](https://github.com/PrestaShop/PrestaShop/pull/38251#issuecomment-2753778184)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AF5GL32I3MI37K2XLNWL2UT2WJSITAVCNFSM6AAAAABZAD5XMWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONJTG43TQMJYGQ> . You are receiving this because you were mentioned.Message ID: ***@***.***>

Pienamente d'accordo ! 😀

kpodemski reviewed (approved):
Well done 👏🏻 

@Codencode this requires modification in the DB, but the modification that is safe to perform and fixes a performance issue, so it seems the PR is acceptable as is :) The last step would be to create a PR against the `autoupgrade`, probably for the `7.0.x` branch, but I will check with the team

jolelievre commented:
@Codencode by the way, the improvement should work without the update on the DB? It's just more efficient with the column indexed right?

Hlavtox commented:
@jolelievre @kpodemski It still brings a massive performance boost even without the index. When I tried all of those queries, I tried it without an index. That's just a bonus.

kpodemski commented:
@Codencode you can use the dev branch of `autoupgrade` to add an index to version `8.2.2`

Codencode commented:
@kpodemski, I created the PR to add the index to the autoupgrade module:
https://github.com/PrestaShop/autoupgrade/pull/1272
Could you review it?
Thanks

Betafer commented:
Hello, 
this is my test with the query suggested on issue: #38095

I have also tested queries on a DB with 9000+ products and 11 languages, prestashop 8.2.1 10.11.11-MariaDB-0+deb12u1

original query takes: 0.0150 s
first query of @Hlavtox: 0.0120 s
second query of @Hlavtox: 0.0029 s
**query of @Eolia : 0.0020 s**

I think this is a good solution, thanks all!

kpodemski commented:
Hello @PrestaShop/qa-functional 👋🏻 

Is there anything else required to test this PR?

Hlavtox commented:
Ping @PrestaShop/committers @PrestaShop/qa-functional

Can I merge this?

Codencode commented:
Hello @jolelievre,
This change requires an update to the autoupgrade module, which I’ve already implemented in this PR: [#1272](https://github.com/PrestaShop/autoupgrade/pull/1272).
I’m sure you’ve noticed, but I just wanted to drop you a quick reminder.
Thank you 🙏

Quetzacoalt91 commented:
Your PR on Update Assistant is not forgotten @Codencode :)
We only wait for this one on the Core to be merged before proceeding in changing the upgrade scripts.

Codencode commented:
Hello @Quetzacoalt91,
great, I just wanted to make sure we didn't overlook this!
Thanks a lot for clarifying!

jolelievre commented:
I just tested the PR by witching the taxes on more than half the products The next product creation correctly switches to the new most used tax by default It works, and automated tests are ok!
